onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -107,26 +108,79 @@ Status AllReduce::ComputeInternal(OpKernelContext* context) const {
 
 AllGather::AllGather(const OpKernelInfo& info) : NcclKernel(info) {
   info.GetAttrOrDefault("group_size", &group_size_, static_cast<int64_t>(1));
+  info.GetAttrOrDefault("axis", &axis_, static_cast<int64_t>(0));
+  cuda_ep_ = static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider());
 }
 
 Status AllGather::ComputeInternal(OpKernelContext* context) const {
   ncclComm_t comm = nccl_->Comm();
 
   auto input_tensor = context->Input<Tensor>(0);
   const void* input_data = input_tensor->DataRaw();
-  const auto in_shape = input_tensor->Shape();
+  const auto& in_shape = input_tensor->Shape();
   int64_t input_count = in_shape.Size();
-  // construct output shape
-  TensorShape out_shape(in_shape);
-  out_shape[0] = group_size_ * out_shape[0];
 
-  void* output_data = context->Output(0, out_shape)->MutableDataRaw();
+  if (axis_ > 0) {
+    // Need transpose
```

Comment:
nit: This is an opportunity for future communication optimization by fusing Transpose and AllGather.

Reply:
we actually can even further fuse allgather with gemm :)